src/tool/xmeta/cpp/models/namespace_model.cpp

```diff
@@ -0,0 +1,120 @@
+    {
+        m_enums.insert(std::pair(std::string_view{ em->get_id() }, em));
+    }
+
+    void namespace_body_model::add_delegate(std::shared_ptr<delegate_model> const& dm)
+    {
+        m_delegates.insert(std::pair(std::string_view{ dm->get_id() }, dm));
+    }
+
+    void namespace_body_model::add_using_alias_directive(std::shared_ptr<using_alias_directive_model> const& uad)
+    {
+        m_using_alias_directives.insert(std::pair(std::string_view{ uad->get_id() }, uad));
+    }
+
+    void namespace_body_model::add_using_namespace_directive(std::shared_ptr<using_namespace_directive_model> const& und)
+    {
+        m_using_namespace_directives.emplace_back(und);
+    }
+
+    // Checks if two member IDs are semantically the same.
+    // namespace member IDs are case preserving but insensitive. This means if 'classA' is defined,
+    // you cannot define another member 'cLaSsA'. You also cannot reference 'classA' using 'cLaSsA'
+    // since case is preserved.
+    template<typename T>
+    struct same_member_id
+    {
+        same_member_id(std::string_view const& name) : new_name{ name } { }
+        bool operator()(std::pair<std::string_view, std::shared_ptr<T>> const& v) const
+        {
+            auto old_name = v.first;
+#if defined(_WIN32)
+            return (stricmp(old_name.data(), new_name.data()) == 0 && new_name != old_name) ||
+                new_name == old_name;
+#endif
+            return false; // Only works on windows for now.
+        }
+    private:
+        std::string_view new_name;
+    };
+
+    inline bool namespace_body_model::type_id_exists(std::string_view const& member_id) const
+    {
+        return contains_id(m_classes, member_id) ||
+            contains_id(m_structs, member_id) ||
+            contains_id(m_interfaces, member_id) ||
+            contains_id(m_enums, member_id) ||
+            contains_id(m_delegates, member_id);
+    }
+
+    void namespace_model::add_child_namespace(std::shared_ptr<namespace_model> const& child)
+    {
+        assert(m_child_namespaces.find(child->get_id()) == m_child_namespaces.end());
+        m_child_namespaces[child->get_id()] = child;
+    }
+
+    void namespace_model::add_namespace_body(std::shared_ptr<namespace_body_model> const& body)
+    {
+        m_namespace_bodies.emplace_back(body);
+    }
+
+    bool namespace_model::member_id_exists(std::string_view const& member_id) const
+    {
+        for (auto ns_body : m_namespace_bodies)
+        {
+            if (ns_body->type_id_exists(member_id))
+            {
+                return true;
+            }
+        }
+
+        return contains_id(m_child_namespaces, member_id);
+    }
+
+    bool namespace_model::child_namespace_exists(std::string_view const& member_id) const
+    {
+        return m_child_namespaces.find(member_id) != m_child_namespaces.end();
+    }
+
+    std::string namespace_model::get_full_name() const
```

Comment:
It's almost guaranteed that this will be asked for. Seems like we should just store this in a member variable and initialize it in the constructor.